### ISSUE TITLE
Docs: [no-unnecessary-type-parameters] add FAQ for some tricky cases

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9680]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

NUTP is an incredibly subtle rule - huge shoutout to @danvk and @JoshuaKGoldberg for their work to implement it!

From testing out this rule, some of the cases that it reports are unusually tricky to fix, compared to most lint rules. In particular, while fixing the reported function signature is generally straightforward, it can have unexpected impacts at the usage sites.

Let's document some of these subtleties on the rule page, to help users out. 

Below is one case that I've had trouble with... I intend to update this post with a few more when I get around to it. I would also welcome anyone else to document tricky cases here!

<details>
<summary>Excess property errors at call sites</summary>
In the following example, the generic property appears unnecessary, and is flagged by the rule.

```ts
interface SomeProperties {
   requiredProperty1: string;
   requiredProperty2: number;
}

function serialize<T extends SomeProperties>(x: T): string {
//                 ~~~~~~~~~~~~~~~~~~~~~~~~
    return JSON.stringify(x);
}
```

The obvious fix would be to replace the generic with its constraint, like so 
```ts
function serialize(x: SomeProperties): string {
    return JSON.stringify(x);
}
```

However, this is a (slightly) breaking change for consumers! (affecting the type system only, not runtime behavior)

Before the change, this usage is just fine:
```ts
serialize({ 
   requiredProperty1: 'foo',
   requiredProperty2: 123,
   excessProperty: 'no problem'
});
```

But afterwards, we get a TS error that `excessProperty` does not exist on type `SomeProperties` (this happens specifically because we're supplying an object literal as the argument; a variable won't trigger this check). And, often, that's the right behavior! There is usually no need for excess properties.

But, sometimes (as, possibly, in the example given here), we actually did want to allow excess properties, and perhaps that's even why the original author wrote the function that way. 

In order to preserve that behavior, the fix is to reflect reasonable excess properties in the index signature of `SomeProperties`. Changing it to 

```ts
interface SomeProperties {
   requiredProperty1: string;
   requiredProperty2: number;
   [anyOtherProperties: string | number | symbol]: unknown;
}
```

we get our original, permissive behavior back. 

Deciding whether to fix the function signature type or the call sites is a case-by-case matter.
</details>

### Affected URL(s)

https://typescript-eslint.io/rules/no-unnecessary-type-parameters/#FAQ

**Comment by danvk** [comment]
Interesting example. I agree that NUTP fixes can sometimes be subtle. It's almost like a puzzle figuring out what the fix is. But there is almost always a fix!

Your example is interesting, and not one I'd thought about. Excess property checking is kind of part of the type system and kind of not, and in retrospect it's not surprising that the fixes for this rule would affect it. I like your workaround of adding the index signature when you explicitly want to allow excess properties. The one think I'd change is `string | number | symbol` → `PropertyKey`, which is an alias for the same thing. I think this would be good to call out in the docs for this rule.

**Comment by kirkwaiblinger** [comment]
Explaining `NoInfer` would be another one to add to the FAQ, since it's an extremely niche feature, see https://github.com/typescript-eslint/typescript-eslint/issues/9751

@bradzacher
